File: Samples/AdventureWorksModel/Mapping/ProductVendorMap.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Data.Entity.ModelConfiguration;

namespace AdventureWorksModel
{
    public class ProductVendorMap : EntityTypeConfiguration<ProductVendor>
    {
        public ProductVendorMap()
        {
            // Primary Key
            HasKey(t => new { t.ProductID, t.VendorID });

            // Properties
            Property(t => t.ProductID)
                .HasDatabaseGeneratedOption(DatabaseGeneratedOption.None);

            Property(t => t.VendorID)
                .HasDatabaseGeneratedOption(DatabaseGeneratedOption.None);

            Property(t => t.UnitMeasureCode)
                .IsRequired()
                .IsFixedLength()
                .HasMaxLength(3);

            // Table & Column Mappings
            ToTable("ProductVendor", "Purchasing");
            Property(t => t.ProductID).HasColumnName("ProductID");
            Property(t => t.VendorID).HasColumnName("BusinessEntityID");
            Property(t => t.AverageLeadTime).HasColumnName("AverageLeadTime");
            Property(t => t.StandardPrice).HasColumnName("StandardPrice");
            Property(t => t.LastReceiptCost).HasColumnName("LastReceiptCost");
            Property(t => t.LastReceiptDate).HasColumnName("LastReceiptDate");
            Property(t => t.MinOrderQty).HasColumnName("MinOrderQty");
            Property(t => t.MaxOrderQty).HasColumnName("MaxOrderQty");
            Property(t => t.OnOrderQty).HasColumnName("OnOrderQty");
            Property(t => t.UnitMeasureCode).HasColumnName("UnitMeasureCode");
            Property(t => t.ModifiedDate).HasColumnName("ModifiedDate").IsConcurrencyToken(false);

            // Relationships
            HasRequired(t => t.Product).WithMany().HasForeignKey(t => t.ProductID);
            HasRequired(t => t.UnitMeasure).WithMany().HasForeignKey(t => t.UnitMeasureCode);
            HasRequired(t => t.Vendor)
                .WithMany(t => t.Products)
                .HasForeignKey(d => d.VendorID);

        }
    }
}

File: Test/AdventureWorksLegacyModelVB/Mapping/ProductVendorMap.vb
Imports System.ComponentModel.DataAnnotations.Schema
Imports System.Data.Entity.ModelConfiguration

Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Metadata.Builders

Namespace AW.Mapping
	Public Class ProductVendorMap
		Inherits EntityTypeConfiguration(Of ProductVendor)

		Public Sub New()
			' Primary Key
			HasKey(Function(t) New With {
				Key t.ProductID,
				Key t.VendorID
			})

			' Properties
			[Property](Function(t) t.ProductID).HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)

			[Property](Function(t) t.VendorID).HasDatabaseGeneratedOption(DatabaseGeneratedOption.None)

			[Property](Function(t) t.UnitMeasureCode).IsRequired().IsFixedLength().HasMaxLength(3)

			' Table & Column Mappings
			ToTable("ProductVendor", "Purchasing")
			[Property](Function(t) t.ProductID).HasColumnName("ProductID")
			[Property](Function(t) t.VendorID).HasColumnName("BusinessEntityID")
			[Property](Function(t) t.AverageLeadTime).HasColumnName("AverageLeadTime")
			[Property](Function(t) t.StandardPrice).HasColumnName("StandardPrice")
			[Property](Function(t) t.LastReceiptCost).HasColumnName("LastReceiptCost")
			[Property](Function(t) t.LastReceiptDate).HasColumnName("LastReceiptDate")
			[Property](Function(t) t.MinOrderQty).HasColumnName("MinOrderQty")
			[Property](Function(t) t.MaxOrderQty).HasColumnName("MaxOrderQty")
			[Property](Function(t) t.OnOrderQty).HasColumnName("OnOrderQty")
			[Property](Function(t) t.UnitMeasureCode).HasColumnName("UnitMeasureCode")
			[Property](Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			HasRequired(Function(t) t.Product).WithMany().HasForeignKey(Function(t) t.ProductID)
			HasRequired(Function(t) t.UnitMeasure).WithMany().HasForeignKey(Function(t) t.UnitMeasureCode)
			HasRequired(Function(t) t.Vendor).WithMany(Function(t) t.Products).HasForeignKey(Function(d) d.VendorID)
		End Sub
	End Class

	Public Module Mapper
		<System.Runtime.CompilerServices.Extension> _
		Public Sub Map(ByVal builder As EntityTypeBuilder(Of ProductVendor))
			' Primary Key
			builder.HasKey(Function(t) New With {
				Key t.ProductID,
				Key t.VendorID
			})

			' Properties
			builder.Property(Function(t) t.ProductID).ValueGeneratedNever()

			builder.Property(Function(t) t.VendorID).ValueGeneratedNever()

			builder.Property(Function(t) t.UnitMeasureCode).IsRequired().IsFixedLength().HasMaxLength(3)

			' Table & Column Mappings
			builder.ToTable("ProductVendor", "Purchasing")
			builder.Property(Function(t) t.ProductID).HasColumnName("ProductID")
			builder.Property(Function(t) t.VendorID).HasColumnName("BusinessEntityID")
			builder.Property(Function(t) t.AverageLeadTime).HasColumnName("AverageLeadTime")
			builder.Property(Function(t) t.StandardPrice).HasColumnName("StandardPrice")
			builder.Property(Function(t) t.LastReceiptCost).HasColumnName("LastReceiptCost")
			builder.Property(Function(t) t.LastReceiptDate).HasColumnName("LastReceiptDate")
			builder.Property(Function(t) t.MinOrderQty).HasColumnName("MinOrderQty")
			builder.Property(Function(t) t.MaxOrderQty).HasColumnName("MaxOrderQty")
			builder.Property(Function(t) t.OnOrderQty).HasColumnName("OnOrderQty")
			builder.Property(Function(t) t.UnitMeasureCode).HasColumnName("UnitMeasureCode")
			builder.Property(Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			builder.HasOne(Function(t) t.Product).WithMany().HasForeignKey(Function(t) t.ProductID)
			builder.HasOne(Function(t) t.UnitMeasure).WithMany().HasForeignKey(Function(t) t.UnitMeasureCode)
			builder.HasOne(Function(t) t.Vendor).WithMany(Function(t) t.Products).HasForeignKey(Function(d) d.VendorID)
		End Sub
	End Module
End Namespace